ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveryCustodyTest.java

```diff
@@ -289,35 +269,86 @@ public void shouldWaitUntilTimeoutedAndHalfOfSidecars() {
-                  Optional.ofNullable(sidecars.get(id.columnIndex())));
-            });
     when(miscHelpersFulu.reconstructAllDataColumnSidecars(anyCollection(), any()))
         .thenReturn(sidecars.values().stream().toList());
     stubAsyncRunner.executeDueActionsRepeatedly();
     stubTimeProvider.advanceTimeBySeconds(1);
     stubAsyncRunner.executeDueActionsRepeatedly();
 
-    verify(delegate, never()).getCustodyDataColumnSidecar(any());
-
     stubTimeProvider.advanceTimeBySeconds(1);
     stubAsyncRunner.executeDueActionsRepeatedly();
 
-    verify(delegate, never()).getCustodyDataColumnSidecar(any());
-
     custody.onNewValidatedDataColumnSidecar(sidecars.get(UInt64.ZERO));
 
     stubAsyncRunner.executeDueActionsRepeatedly();
-    // prepare
-    verify(delegate, times(64)).getCustodyDataColumnSidecar(any());
+
+    verify(miscHelpersFulu).reconstructAllDataColumnSidecars(anyCollection(), any());
 
     // post reconstructed
     verify(delegate, times(config.getNumberOfColumns())).onNewValidatedDataColumnSidecar(any());
     verify(listener, times(64)).onNewValidSidecar(any(), eq(RemoteOrigin.RECOVERED));
     verify(dataColumnSidecarPublisher, times(64)).accept(any());
   }
+
+  @Test
+  public void shouldRestartOnException() {
+    custody.onSlot(slot);
+    assertThat(stubAsyncRunner.hasDelayedActions()).isTrue();
+
+    custody.onNewBlock(signedBeaconBlock, Optional.empty());
+    final Map<UInt64, DataColumnSidecar> sidecars =
+        columnIndices
+            .get()
+            .map(i -> dataStructureUtil.randomDataColumnSidecar(signedBeaconBlock.asHeader(), i))
+            .collect(Collectors.toMap(DataColumnSidecar::getIndex, sidecar -> sidecar));
+    sidecars.values().stream().skip(30).limit(70).forEach(custody::onNewValidatedDataColumnSidecar);
+
+    when(miscHelpersFulu.reconstructAllDataColumnSidecars(anyCollection(), any()))
+        .thenThrow(new RuntimeException("Simulated exception"));
+    stubAsyncRunner.executeDueActionsRepeatedly();
+    stubTimeProvider.advanceTimeBySeconds(2);
+    stubAsyncRunner.executeDueActionsRepeatedly();
+
+    verify(miscHelpersFulu).reconstructAllDataColumnSidecars(anyCollection(), any());
+
+    verify(listener, never()).onNewValidSidecar(any(), eq(RemoteOrigin.RECOVERED));
+    verify(dataColumnSidecarPublisher, never()).accept(any());
+
+    // add 1 sidecar to restart
+    reset(miscHelpersFulu);
+    when(miscHelpersFulu.reconstructAllDataColumnSidecars(anyCollection(), any()))
+        .thenReturn(sidecars.values().stream().toList());
+    custody.onNewValidatedDataColumnSidecar(sidecars.get(UInt64.ZERO));
+    stubAsyncRunner.executeDueActionsRepeatedly();
+
+    // post reconstructed
+    verify(delegate, times(config.getNumberOfColumns())).onNewValidatedDataColumnSidecar(any());
+    verify(listener, times(57)).onNewValidSidecar(any(), eq(RemoteOrigin.RECOVERED));
+    verify(dataColumnSidecarPublisher, times(57)).accept(any());
+  }
```

Comment:
do we really need 1 test that does all that?

this looks like we're testing multiple things and putting it all in 1 test. can it be split out to the separate things its testing?